src/main/java/frc/robot/subsystems/vision/VisionEstimation.java

```diff
@@ -0,0 +1,159 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package frc.robot.subsystems.vision;
+
+import static edu.wpi.first.apriltag.AprilTagFieldLayout.OriginPosition.kBlueAllianceWallRightSide;
+import static edu.wpi.first.apriltag.AprilTagFieldLayout.OriginPosition.kRedAllianceWallRightSide;
+
+import org.photonvision.EstimatedRobotPose;
+import org.photonvision.PhotonCamera;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout.OriginPosition;
+import edu.wpi.first.math.Matrix;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.numbers.N1;
+import edu.wpi.first.math.numbers.N3;
+import edu.wpi.first.wpilibj.Notifier;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants.FieldConstants;
+import frc.robot.Constants.VisionConstants;
+import frc.robot.subsystems.swervedrive.SwerveSubsystem;
+
+public class VisionEstimation extends SubsystemBase {
+    private final SwerveSubsystem swerveSubsystem;
+
+    public Matrix<N3, N1> visionMeasurementStdDevs = VecBuilder.fill(0.9,
+            0.9,
+            0.9);
+
+    private final Vision frontRightEstimator = new Vision(new PhotonCamera("Inky"),
+            VisionConstants.ROBOT_TO_INKY);
+    private final Vision frontLeftEstimator = new Vision(new PhotonCamera("Blinky"),
+            VisionConstants.ROBOT_TO_BLINKY);
+    private final Vision backRightEstimator = new Vision(new PhotonCamera("Pinky"),
+            VisionConstants.ROBOT_TO_PINKY);
+    private final Vision backLeftEstimator = new Vision(new PhotonCamera("Clyde"),
+            VisionConstants.ROBOT_TO_CLYDE);
+
+    private final Notifier allNotifier = new Notifier(() -> {
+        frontRightEstimator.run();
+        frontLeftEstimator.run();
+        backRightEstimator.run();
+        backLeftEstimator.run();
+    });
+
+    private OriginPosition originPosition = kBlueAllianceWallRightSide;
+
+    public VisionEstimation() {
+        this.swerveSubsystem = SwerveSubsystem.getInstance();
+
+        allNotifier.setName("runAll");
+        allNotifier.startPeriodic(0.02);
+
+        // prevTime.set(Timer.getFPGATimestamp());
+        this.swerveSubsystem.getPose();
+    }
+
+    @Override
+    public void periodic() {
+        if (VisionConstants.USE_VISION) {
+            estimatorChecker(frontRightEstimator);
+            estimatorChecker(frontLeftEstimator);
+            estimatorChecker(backRightEstimator);
+            estimatorChecker(backLeftEstimator);
+        } else {
+            allNotifier.close();
+        }
```

Comment:
If `USE_VISION` is false, we would be calling `allNotifier.close` on every periodic event. We could just check USE_VISION in the constructor and if it is false, we don't need to call `allNotifier.startPeriodic` and we won't need to call `allNotifier.close` either.

Reply:
We need to call startPeriodic and with it checking in periodic can just disable vision in the middle on the game if needed

Reply:
If we plan to enable/disable Vision during the game, then this code is not handling the disabled->enabled transition. It only handles the enabled->disabled transition. If we want to truly support dynamic turning on/off the Vision, we need some more code...

Reply:
Well we haven't added the code now but can add it later to control it if we actually need to